### DISCUSSION TITLE
Mehrere aufeinanderfolgende Builds führen zu unterschiedlichen Images

### DISCUSSION THREAD

**Discussion by BugExplorer-Plus** [Freetz-NG/freetz-ng#881] [Allgemein] [answered]
Habe die Frage schon bei hoerli.net (https://hoerli.net/freetz-ng-hol-mehr-aus-deiner-fritzbox/) platziert, denn nach dieser Anleizung habe ich die Builds durchgeführt:
Mehrere Builds führen zu sehr unterschiedlichen binär-Images, auch wenn der make im Minutentakt ausgeführt wird – Die Dateigröße ist identisch (41.308.160 Bytes/build für 7530ax/FW ab 7.50), die Images unterscheiden sich bis zur Position 1178, danach geht es bis 3536989 synchron weiter und dann ist fast bis zum Ende (ca. 36 MB) alles ungleich, eine Übereinstimmung durch Verschiebung lässt sich nicht erkennen. Der zweite Bereich ab dem die Unterschiede auftauchen wird eingeleitet von “brcma9_HW256_squashFS_filesystem”, handelt es sich hierbei um ein nicht-determistisches Schreiben in ein Dateisystem???
Diese großen Unterschiede empfinde ich zunächst als merkwürdig, fällt jemand etwas dazu ein?

**Comment by fda77** [comment, marked as answer]
Der erste Teil ist der Kernel. Den ersetzt du nicht da es das in Freetz für dies Gerät nicht gibt sondern nimmst den unverändert von AVM. Der Rest ist das Dateisystem das *komprimiert* ist. Allein schon durch 1 anderes Datum 1 Datei oder Verzeichnis ändert sich das ganze.
Auch ist in Dateien ein Datum drin, zb zeigt Busybox beim einloggen an wann sie gebaut wurde 
Gibts bei Freetz auch nicht: https://en.wikipedia.org/wiki/Reproducible_builds

**Reply by BugExplorer-Plus** [reply]
Danke für die schnelle Antwot - ich verstehe jetzt - für mich war der "Reproducible build" eins der Kriterien für die Korrektheit der Build Prozedur. Gibt es eine andere Möglichkeit - ohne Flashen - den Build zu validieren?

**Reply by fda77** [reply]
Also ich mach das mangels Hardware immer nach Gefühl. Fall doch was nicht geht wird sich schon jemand melden

**Reply by kainers1** [reply]
Du müsstest das angesprochene Dateisystem zuerst auspacken und dann vergleichen

**Reply by BugExplorer-Plus** [reply]
Klingt interessant - man müsste also:
- das Dateisystem klar abgrenzen
- mittels der squashfs-tools komplett expandieren
- beide Images bearbeiten/expandieren und mit einem hierarchisch/rekusiven Vergleich die Differenz(en) bestimmen und
- für jede Abweichung eine schlüssige Erklärung finden, die die Abweichung ableitbar macht

**Comment by fda77** [comment]
Nur zu, wenn du nichts besseres zu tun hast...

 - Ich würde zuerst 2x freetz auschecken und je mit gleicher .config + revision ein image bauen - es könnten die compiler etc host-pfade in binaries sein.
 - diese mit tools/fwdu entpacken
 - md5 aller dateien mit find erstellen und vergleichen
 - für jede lib+package die es betrifft eine option hinzufügen die es reproducible macht (für busybox gäbe es das schon: https://github.com/Freetz-NG/freetz-ng/blob/master/make/busybox/Config.in#L564)
 - in menuconfig einen hauptschalter "reproducible" machen

Das wäre mir zuviel Aufwand für keinen Nutzen... ausser du hast auch mit den Pay patches+images im DEB zu tun (https://www.ip-phone-forum.de/threads/316603/)

**Reply by Prisrak1** [reply]
Also bitte, Ich hab noch nie irgendwas von jemanden für meine Leistungen irgendwann mal auch einen Cent bekommen oder verlangt. Nur weil ich irgendwas aus meiner Unwissenheit eben irgendwo etwas erfrage, gibt es dahinter auch keinen Hintergrund. Für mich ist das überhaupt nicht nachvollziehbar, weshalb bist du so negativ zu mir und zu unserem Forum so eingestellt bist. Vermutlich ist von dir überhaupt nicht nachvollziehbar, dass es so manche Menschen gibt, die irgendwo sich für kein Geld und somit kostenlos sich engagieren, um jemanden einfach so manche Tatsachen etwas plausibler näherzubringen versuchen. Man muss schon deine niedergeschriebene Aussagen schon irgendwie richtig bedauern.

**Reply by BugExplorer-Plus** [reply]
Hmmm...

**Reply by BugExplorer-Plus** [reply]
Also mit der Kommerzialisierung von freetz habe ich nichts zu tun - meine Einstellung zu diesem Forum ist ebenfalls nicht negativ, denn sonst hätte ich erst gar nicht hier gefragt.
Dass die Frage auch neutral/sachlich aufgefasst werden kann, zeigen auch die Ausführungen von fda77 und berndy2001!

**Reply by fda77** [reply]
@Prisrak1 ich find es halt nur sehr auffällig dass du im IPPF nach dem Geld fragst was man fürs bauen verlangen könnte, während sammy neuerdings patches für freetz nur noch gegen Geld via www.spendenseite.de anbietet und dann hier ein Neuer nach Überprüfbarkeit gebauter Images fragt. Das kommt mit halt etwas spanisch vor.
Und ja, das ist ein grundlegendes Problem mit den DEB Leuten. Denn statt hier beizutragen wird das ganze als Einbahnstrasse angesehen.

**Reply by BugExplorer-Plus** [reply]
@fda77: Also die Begriffe IPPF und DEB sind mir unbekannt - die Suchmaschine liefert mir hierzu nichts passendes, allerdings wäre eine diesbezügliche Erläuterung vielleicht hilfreich, denn diese Diskussion ist mir noch rätselhafter als die Ausgangssituation der "non reproducible images"...

**Reply by BugExplorer-Plus** [reply]
Wir sind wohl übrigens ziemlich weit Off Topic!

**Reply by fda77** [reply]
Dann erzähl halt wie weit du mittlerweile bist...

**Reply by BugExplorer-Plus** [reply]
Ich versuche mich jetzt daran mit Debian11 einen Teil aus einer Datei herauszukopieren - meine umfangreicheren Experimente mit Linux liegen schon ca. 15 Jahre zurück. Wenn ich den richtigen Ausschnitt habe, könnte ich dann mit unquashfs die Verzeichnishierarchie für beide Images erstellen und wie ich dann die md5s über die Verzeichnishierarchie erstelle ist mir auch noch unklar...

**Reply by fda77** [reply]
Du könntest es dir auch einfach machen und so entpacken wie ich oben geschrieben hab

**Reply by BugExplorer-Plus** [reply]
Ich glaube ich habe zum ganzen Thema einfach noch zu wenig Background und muss mich tiefer reinwühlen - danke auf alle Fälle für Deine Unterstützung und Dein sachliches Feedback - sobald ich was konkretes habe, poste ich es hier!

**Reply by fda77** [reply]
Jo, mach mal eine Übersicht welche Pakete man alles anpassen müsste. Vielleicht sind es nicht so viele

**Reply by BugExplorer-Plus** [reply]
Hallo fda77, bei mir entstehen bei dieser Prozedur folgende Fehler:
tools/fwdu verbose unpack images/latest.image
    You have to run 'make tools-all' first.

make tools-all
    ERROR: modpatch: Error in patch-file /home/user/freetz-ng/make/host-tools/openssl-host/patches/170-no-scripts-install.patch

**Reply by fda77** [reply]
Irgend was geht mit openssl-host nicht

**Reply by BugExplorer-Plus** [reply]
Nochmal genauer:
    
    applying patch file /home/user/freetz-ng/make/host-tools/openssl-host/patches/170-no-scripts-install.patch
    patching file Configurations/unix-Makefile.tmpl
    Hunk #1 FAILED at 649.
    1 out of 1 hunk FAILED -- saving rejects to file Configurations/unix-Makefile.tmpl.rej
    ERROR: modpatch: Error in patch-file /home/user/freetz-ng/make/host-tools/openssl-host/patches/170-no-scripts-install.patch

Die Dateien: /home/user/freetz-ng/make/host-tools/openssl-host/patches/170-no-scripts-install.patch
und:  /home/user/freetz-ng/source/host-tools/openssl-3.0.10/Configurations/unix-Makefile.tmpl.rej
sind inhaltsgleich. Ich lade mal die Patch-Datei mit hoch!
[170-no-scripts-install.patch](https://github.com/Freetz-NG/freetz-ng/files/13527435/170-no-scripts-install.patch)

**Reply by fda77** [reply]
Und du bist sicher dass du nicht irgend was verbastelt hast? Dieser Patch ist ein Link auf das package bzw target-openssl: https://github.com/Freetz-NG/freetz-ng/blob/master/make/host-tools/openssl-host/patches/170-no-scripts-install.patch
Und somit wird der Patch für host automatisch mitaktualisiert

**Reply by BugExplorer-Plus** [reply]
Ich wüsste nicht was - ich habe lediglich das Skript Freetz.sh in einem Minimal debian 11 ausgeführt, das dann den make für das Image startete - an der entstandenen Struktur oder den Dateien habe ich keine Änderungen gemacht.

**Reply by BugExplorer-Plus** [reply]
Was mich noch verwirrt: Obwohl die Dateien exakt inhaltsgleich sind, zeigt kate ein anderes syntax-highlighting an...

**Reply by fda77** [reply]
Na dann such mal. Ich mag es übrigens nach wie vor überhaupt nicht wenn man selbst herumbastelt und dann Fehler DIE DADURCH ENTSTEHEN hier meldet. Das verschwendet meine Zeit

```
$ make openssl-host-unpacked

---> tools/openssl-host ...
preparing ... mkdir -p /home/freetz-ng/freetz-ng/source/host-tools/openssl-3.0.12; tools/gunzip -c dl/openssl-3.0.12.tar.gz | tools/tar-gnu -x -C /home/freetz-ng/freetz-ng/source/host-tools/openssl-3.0.12 --transform='s|^./\+||' --strip-components=1
set -e; shopt -s nullglob; for i in /home/freetz-ng/freetz-ng/make/host-tools/openssl-host/patches/*.patch*; do case $i in *.patch|*.patch.gz|*.patch.bzip2|*.patch.bz2|*.patch.bz|*.patch.xz|*.patch.lz|*.patch.lzma|*.patch.Z|*.patch.diff) ;; *) continue ;; esac; tools/freetz_patch /home/freetz-ng/freetz-ng/source/host-tools/openssl-3.0.12 $i ; done;
    applying patch file /home/freetz-ng/freetz-ng/make/host-tools/openssl-host/patches/130-dont-build-fuzz.patch
    patching file build.info
    ----------------------------------------------------------------------
    applying patch file /home/freetz-ng/freetz-ng/make/host-tools/openssl-host/patches/160-disable_docs.openssl.patch
    patching file Configurations/unix-Makefile.tmpl
    ----------------------------------------------------------------------
    applying patch file /home/freetz-ng/freetz-ng/make/host-tools/openssl-host/patches/170-no-scripts-install.patch
    patching file Configurations/unix-Makefile.tmpl
    ----------------------------------------------------------------------
    applying patch file /home/freetz-ng/freetz-ng/make/host-tools/openssl-host/patches/180-no-bins.patch
    patching file build.info
    ----------------------------------------------------------------------
(cd /home/freetz-ng/freetz-ng/source/host-tools/openssl-3.0.12; rm -f Makefile Makefile.bak; ln -s Configure configure;)
 ```

**Comment by Prisrak1** [comment]
".. sammy neuerdings patches für freetz nur noch gegen Geld via [www.spendenseite.de) anbietet.. " Ist mir nicht bekannt und ich würde niemals für einen Patch etwas bezahlen oder Geld nehmen. Er hat nur Probleme mit dem Hochladen auf dem Teamserver. Ich kann es aber nicht beheben. Zudem wenn etwas von mir geladen wird was nicht da rein gehört, nutze ich auch kostenlose Webseiten dazu. Die ganzen Anschuldigungen finde ich nicht gerecht. Ich finde man soll sich immer aussprechen und die Tatsachen untereinander klären. So entstehen die Konflikte. Obwohl es off Topic ist,

**Reply by fda77** [reply]
> Die ganzen Anschuldigungen

Keine Ahnung was (welche) du meinst, das sind Tatsachen zu den ich mir das wahrscheinliste denke. Solang du aber nicht konkret wirst ist das eh nr blah und man kann dazu nicht wirklich was schreiben

**Comment by BugExplorer-Plus** [comment]
Auf alle Fälle kann ich schon mal mein komplette Build-Script zur Begutachtung vorlegen -vielleicht habe ich da noch ein Problem eingebaut, allerdings ist die einzige Fehlermeldung, dass der Zugriff auf /dev/tty verweigert wurde.
[Freetz.sh.txt](https://github.com/Freetz-NG/freetz-ng/files/13267072/Freetz.sh.txt)

**Reply by fda77** [reply]
execstack ist mittlerweile ein hosttool (prelink) und wget ein busybox applet. Das mit den Pfaden für Debian hat auch einen workaround.
Das steht soweit alles schon hier https://github.com/Freetz-NG/freetz-ng#readme , prerequisites könnte man noch direkt hinschreiben

**Reply by fda77** [reply]
Das mit dem tty tritt übrigens nur auf gewissen verkonfigurierten uralt dell servern auf ;-)

**Comment by fraternl** [comment]
@BugExplorer-Plus 

You mention the 7530AX.
Do you mean you can successfully flash a 7530AX with tools/push_firmware ??

I can't.
I am able to compile it, but when I flash it, it comes back with standard firmware.
I could try with a fully compiled freetz-ng image to see if it's the image or the environment.

If I succeed to get Freetz-NG on it, I can, from then on flash it with Freetz-NG through the WebIF
